app/src/main/kotlin/compiler/intermediate_form/Allocation.kt

```diff
@@ -1,14 +1,105 @@
 package compiler.intermediate_form
 
+import java.util.LinkedList
+import kotlin.collections.HashMap
+
 object Allocation {
     data class AllocationResult(
         val allocatedRegisters: Map<Register, Register>,
         val spilledRegisters: List<Register>,
     )
 
+    private fun Map<Register, Set<Register>>.sortedByRemovingNodesWithSmallestDeg(): List<Register> = mutableListOf<Register>().apply {
+        val graph = this@sortedByRemovingNodesWithSmallestDeg.mapValues { it.value.toHashSet() }.toMutableMap()
+
+        fun removeFromGraph(node: Register) {
+            graph[node]!!.forEach { graph[it]!!.remove(node) }
+            graph.remove(node)
+        }
+
+        fun getNodeWithSmallestDegree(): Register = graph.minByOrNull { it.value.size }!!.key
+
+        while (graph.isNotEmpty())
+            getNodeWithSmallestDegree().let {
+                removeFromGraph(it)
+                add(it)
+            }
+    }
+
+    private class GraphColoring(
+        val graph: Map<Register, Set<Register>>,
+        val accessibleRegisters: List<Register>
+    ) {
+        private val allocationMap = HashMap<Register, Register>().apply { putAll(accessibleRegisters.associateWith { it }) }
+        private val availableColorsMap = HashMap<Register, HashSet<Register>>()
+        private val spilledRegisters = LinkedList<Register>()
+        fun Register.isColored() = allocationMap.contains(this)
+        fun Register.color() = allocationMap[this]
+        fun Register.availableColors() = availableColorsMap.getOrPut(this) { accessibleRegisters.toHashSet() }
+        fun Register.assignColor(color: Register) {
+            allocationMap[this] = color
+            graph[this]!!.forEach { availableColorsMap.getOrPut(it) { accessibleRegisters.toHashSet() }.remove(color) }
+        }
+
+        fun Register.spill() = spilledRegisters.add(this)
+        fun createAllocationResult() = AllocationResult(allocationMap, spilledRegisters)
+    }
+
     fun allocateRegisters(
         linearProgram: List<Asmable>,
         livenessGraphs: Liveness.LivenessGraphs,
         accessibleRegisters: List<Register>,
-    ): AllocationResult = TODO()
+    ): AllocationResult = GraphColoring(livenessGraphs.interferenceGraph, accessibleRegisters).run {
+        val copyGraphWithoutInterferences = livenessGraphs.copyGraph.mapValues { it.value - livenessGraphs.interferenceGraph[it.key]!! }
+
+        for (register in livenessGraphs.interferenceGraph.sortedByRemovingNodesWithSmallestDeg().reversed()) {
+
+            val bestForColored: Register? by lazy(LazyThreadSafetyMode.NONE) {
+                findBestFitForColoredCopyGraphNeighbours(register, copyGraphWithoutInterferences)
+            }
+            val bestForUncolored: Register? by lazy(LazyThreadSafetyMode.NONE) {
+                findBestFitForUncoloredCopyGraphNeighbours(register, copyGraphWithoutInterferences)
+            }
+
+            when {
+                register.isColored() -> {}
```

Comment:
Is this even possible?

Reply:
Yes, it is - the interference graph may contain hardware registers as its nodes, which are pre-colored with themselves